quinn-proto/src/connection/mod.rs

```diff
@@ -695,11 +695,16 @@ impl Connection {
                 let next_datagram_size_limit = match self.spaces[space_id].loss_probes {
                     0 => segment_size,
                     _ => {
-                        self.spaces[space_id].loss_probes -= 1;
+                        if num_datagrams >= 1 && segment_size > usize::from(INITIAL_MTU) {
+                            // When building a GSO batch we can not reduce the size of a
+                            // datagram for a tail-loss probe.
+                            break;
```

Comment:
We can and do:

https://github.com/quinn-rs/quinn/blob/8d6e48c20b71f7e915c13c33b66e2a09b6d59888/quinn-proto/src/connection/mod.rs#L639-L657

The trailing packet in a GSO batch may be shorter than the others, so we terminate a GSO batch *after* producing such a packet.

Reply:
Oh right, and `buf_capacity` gets set from the `next_datagram_size` which this PR is fixing.  The whole flow of poll_transmit is still rather complex to me, but it's improving :)

I removed the early break here again.